### PR TITLE
Add new C++ and Rust templates for GUI applications written in Slint

### PR DESCRIPTION
**Pull request by tronical**

### CODE REVIEW

**Comment by tronical** [comment]
I've tried this on a imx8 on a dahlia board and it works for me. The docker file is a little complex, and it also takes a fair while to build. The upside though is that since the slint building part is the same for the SDK, the deploy image for the SDK, as well as the production, once it's built those layers are reused.

**Comment by tronical** [comment]
Regarding the logo: Do you know by chance if SVGs are supported, too, by your extension? That would look better than the PNG file for the Slint logo.

Edit: I changed it to use an SVG - works fine and looks much better :)

**Comment by tronical** [comment]
I've taken the liberty of also adding the Rust template into the same PR.

Happy to make any kind of squashes/rebases/etc. as requested. I can also split it again into tow PRs if you'd like.

**Comment by microhobby** [comment]
Hey @tronical thanks for the contributions! I will take a look soon.

**Comment by microhobby** [comment]
@tronical sorry I only see this now, but all the new development should be merged into the `dev` branch. Could you please use the `dev` branch as base for your development?

Tip: I think that will be easier to you get your commits and cherry-pick to a fresh rebase from dev than try to rebase `dev` into `bookworm`

**Comment by tronical** [comment]
Sure, I'll rebase onto dev.

**Comment by tronical** [comment]
Done. There were no conflicts. Did dev diverge from bookworm in a way that would require making changes to the docker files, etc?

**Comment by microhobby** [comment]
@tronical oh cool, not differ in content, only in commit hashes, so I thought that this will be an issue. But good to know. Thanks!

File: cppSlint/.conf/deps.json
Comment (microhobby) [inline cppSlint/.conf/deps.json]: This package is only needed from Qt because we do some port redirect to use the Qt Creator as debugger. Are you using it for something in the Slint template?

File: cppSlint/.vscode/c_cpp_properties.json
Comment (microhobby) [inline cppSlint/.vscode/c_cpp_properties.json:7]: There is some way to download and configure here the Slint headers, like the `appwindow.h` for example?

Comment (microhobby) [reply cppSlint/.vscode/c_cpp_properties.json:7]: The cpp linter works locally, and we should have it locally to set in the `includePath`.
Maybe to add a task entry to be executed in the project open to download the headers...

File: cppSlint/Dockerfile.sdk
Comment (microhobby) [inline cppSlint/Dockerfile.sdk]: `libstdc++-10-dev` has no candidate for Debian `bookworm`. Is possible to use the `libstdc++-11-dev`? https://packages.debian.org/search?suite=bookworm&section=all&arch=any&searchon=names&keywords=libstdc%2B%2B-

File: cppSlint/Dockerfile.sdk
Comment (microhobby) [inline cppSlint/Dockerfile.sdk:106]: yeah, this take some time to build. Maybe should be good to you guys provide some pre-built Docker images to use as base, like `slint/slint-ui` for `arm64v8` and `arm32v7`. This should greatly improve build speed.

File: assets/img/slint-logo-square-light.svg
Comment (microhobby) [inline assets/img/slint-logo-square-light.svg:1]: For maintaining the consistency, I would like to propose to use the follow images:
slintrust
slintcpp

Let me know if this works for you.

**Review by microhobby — CHANGES_REQUESTED** [review]

**Comment by microhobby** [comment]
@tronical for now I only checked the C++ one, next week we will check the Rust template.

File: cppSlint/Dockerfile.sdk
Comment (microhobby) [reply cppSlint/Dockerfile.sdk:106]: The comment above is only a suggestion, this will not block the merge.

File: cppSlint/.vscode/settings.json
Comment (microhobby) [inline cppSlint/.vscode/settings.json:18]: Did you do not want to add here the:

```json
  "slint.preview.providedByEditor": true
```

I was testing here and this integrates very well instead of the default preview that uses a fullscreen window

File: cppSlint/.conf/deps.json
Comment (tronical) [reply cppSlint/.conf/deps.json]: Indeed, that's a leftover from copy & paste. Will remove - Thanks :)

File: cppSlint/.vscode/c_cpp_properties.json
Comment (tronical) [reply cppSlint/.vscode/c_cpp_properties.json:7]: Hm, `appwindow.h` is a generated header file (from `appwindow.slint`), so I'm not sure where that could be downloaded from/to. The Slint C++ API headers are inside the docker image and installed into /usr/include.

How do you handle this with other C/C++ libraries?

File: cppSlint/.vscode/settings.json
Comment (tronical) [reply cppSlint/.vscode/settings.json:18]: I'd prefer to stay with the defaults here. The side-by-side works, but in terms of screen real estate that remains left for the code editor, I prefer the separate preview window.

File: cppSlint/Dockerfile.sdk
Comment (tronical) [reply cppSlint/Dockerfile.sdk]: Yes, that's the correct choice, I'll fix that. When I started, the 3-bookwork tag only had libstdc++10-dev, but it seems that meanwhile you moved the tag to a newer Debian snapshot where 10 was replaced with 11.

File: cppSlint/Dockerfile.sdk
Comment (tronical) [reply cppSlint/Dockerfile.sdk:106]: Yeah, it's tempting to do that. But it's also a bit of work to maintain them in parallel your images. When you update your 3-bookworm tag, we need to update our images, too. Another option is to entirely deviate from Torizon, but then we might run into issues where our container creates run-time library dependencies that aren't present in the torizon deployment containers.

File: assets/img/slint-logo-square-light.svg
Comment (tronical) [reply assets/img/slint-logo-square-light.svg:1]: Excellent idea! Let me see if I can integrate those tag images into the .svg, otherwise we can use your png files :)

Comment (tronical) [reply assets/img/slint-logo-square-light.svg:1]: Alright, I can't see a trivial way - so I pushed a commit with your png versions. Thanks again :)

File: cppSlint/.vscode/settings.json
Comment (microhobby) [reply cppSlint/.vscode/settings.json:18]: Is there a way to not show the window in full screen, then? My concern is that it makes users, that uses single screen, confused.

File: rustSlint/src/main.rs
Comment (microhobby) [inline rustSlint/src/main.rs]: I think here should be `Hello Torizon!`

**Review by microhobby — CHANGES_REQUESTED** [review]

File: cppSlint/.vscode/settings.json
Comment (microhobby) [reply cppSlint/.vscode/settings.json:18]: ok, I tried again and it's showing a normal window, resolving also this one

File: cppSlint/.vscode/c_cpp_properties.json
Comment (microhobby) [reply cppSlint/.vscode/c_cpp_properties.json:7]: We usually instruct users to install the reference dev package `yourlib-dev`, or download the headers to have a local reference. As I said, the linter runs locally, I mean outside the SDK container.

But from my tests the `build-debug-amd64-local`, already get the `build/x86_64/_deps/slint-src/api/cpp/include` folder that has the API headers. So, LGTM

**Comment by microhobby** [comment]
@tronical thanks a lot for the changes, I check also the rust one and apart from the small detail here https://github.com/toradex/vscode-torizon-templates/pull/52#discussion_r1188081030, LGTM. As soon as you change this detail we will merge it, thank you.

File: rustSlint/src/main.rs
Comment (tronical) [reply rustSlint/src/main.rs]: Ah yes, well spotted!

**Comment by tronical** [comment]
Thanks for the review! I pushed the fix.

**Pull request merged**
